### ISSUE TITLE
HBM_POOL_PLANNING=0 does not fix the 0x7b1b gather fault when HBM is nearly full

### ISSUE THREAD

**Issue by tdoublep** [torch-spyre/torch-spyre#4063]
Follow-up to #4033.

The reproducer in #4033 passes with `HBM_POOL_PLANNING=0`, which makes the fault look like a pool-planner bug. It is not only that. The same no-op gather still faults with pool planning off, once HBM is nearly full.

| resident HBM | `HBM_POOL_PLANNING` | result |
| --- | --- | --- |
| ~0 | unset (on) | fault |
| ~0 | `0` | pass |
| ~15.6 GiB | `0` | fault |

Removing just the gather makes it pass, with nothing else changed. Two things are needed on top of #4033 to see it with pool planning off: the softmax tail after the gather (a bare gather plus one matmul passes at this occupancy), and nearly-full HBM. The occupancy part is about the allocation pattern rather than the total, see the flags below.

This is what we hit in vLLM via spyre-inference. It already sets `HBM_POOL_PLANNING=0`, and an 8B model still faults on the first decode step, so the flag is not a usable workaround there.

Needs a machine with ~16 GiB of free HBM. On ours 15.0 GiB of ballast passes and 15.25 GiB faults, so you may need to adjust `R_BALLAST_GB` to find the threshold on a larger part.

The script has three knobs, all environment variables:

- `R_GATHER` (default 1) — the thing under test. 1 does the no-op gather, 0 uses `query` directly.
- `R_BALLAST_GB` (default 15.25) — GiB of fp16 allocated up front as 256 MiB chunks, standing in for loading a checkpoint. Fills HBM.
- `R_WEIGHTS` (default 12) — a count, not a size: 12 tensors of 4096x4096 fp16, 32 MiB each, allocated after the ballast. The kernel never reads them. They are only there to recreate the allocation size mix, big chunks followed by a run of medium tensors.

Both memory knobs are needed, because the trigger is the mix rather than the total:

| ballast | weights | resident | result |
| --- | --- | --- | --- |
| 15.25 GiB | 12 | ~15.63 GiB | fault |
| 15.25 GiB | 0 | 15.25 GiB | pass |
| 15.75 GiB | 1 | ~15.78 GiB | pass |

The third row has more bytes resident than the first and still passes.

```
HBM_POOL_PLANNING=0 python repro.py                        # fault
HBM_POOL_PLANNING=0 R_GATHER=0 python repro.py             # pass
R_BALLAST_GB=0 R_WEIGHTS=0 python repro.py                 # fault, the #4033 pool-planning path
```

```python
import os

import torch
import torch_spyre  # noqa: F401
from torch_spyre.streams import synchronize

NH, NKV, D, BLOCK, PAGES = 32, 8, 128, 128, 5
NQ = NH // NKV
BALLAST_GB = float(os.environ.get("R_BALLAST_GB", "15.25"))
WEIGHTS = int(os.environ.get("R_WEIGHTS", "12"))
GATHER = os.environ.get("R_GATHER", "1") == "1"

dev = torch.device("spyre")
torch.manual_seed(0)

# Fill HBM first, the way loading a 15 GiB checkpoint would.
side = int((256 * 1024 * 1024 // 2) ** 0.5)
ballast = [
    torch.zeros(side, side, dtype=torch.float16).to(dev)
    for _ in range(int(BALLAST_GB * 1024 / 256))
]
weights = [
    torch.randn(4096, NH * D, dtype=torch.float16).mul_(0.02).to(dev) for _ in range(WEIGHTS)
]


def kernel(query, query_row_index, k_pages, v_pages, page_index_table, mask_tile):
    # query has one row and query_row_index[:1] selects it, so this gather is a no-op.
    q_rows = query.index_select(0, query_row_index[:1]) if GATHER else query
    q = q_rows.unsqueeze(0).transpose(1, 2).reshape(NKV, NQ, 1, D)
    page_idx = page_index_table[0, 0:1]
    k = k_pages.index_select(0, page_idx).squeeze(0).permute(1, 0, 2).unsqueeze(1)
    v = v_pages.index_select(0, page_idx).squeeze(0).permute(1, 0, 2).unsqueeze(1)
    scores = torch.matmul(q, k.transpose(-2, -1)) * D**-0.5 + mask_tile
    probs = torch.exp(scores - torch.amax(scores, dim=-1, keepdim=True))
    attn = torch.matmul(probs, v) / probs.sum(dim=-1, keepdim=True)
    return attn.reshape(1, NH, 1, D).transpose(1, 2).reshape(1, NH, D)


k_pages = torch.randn(PAGES, BLOCK, NKV, D, dtype=torch.float16).mul_(0.1).to(dev)
v_pages = torch.randn(PAGES, BLOCK, NKV, D, dtype=torch.float16).mul_(0.1).to(dev)
query = torch.randn(1, NH * D, dtype=torch.float16).to(dev).reshape(1, NH, D)
query_row_index = torch.zeros(32, dtype=torch.int32).to(dev)
page_index_table = torch.zeros(1, 32, dtype=torch.int32).to(dev)
mask_tile = torch.zeros(1, BLOCK, dtype=torch.float16).to(dev)

run = torch.compile(kernel, dynamic=False)
out = run(query, query_row_index, k_pages, v_pages, page_index_table, mask_tile)
synchronize()
print("PASS -- no fault", tuple(out.shape))
```

Error:

```
[unspecified] ERRR ras_base.hpp:74 {"action":"information","category":"hardware","code":"0x7b1b","description":"A hardware error was detected in the compute control block response.","message":"Compute CB hardware error detected","name":"RAS::RUNTIMESCHEDULER::ComputeHardwareError","severity":"ERROR","step":"Open ticket","type":"runtime_error"}
```

torch-spyre a31289852145a59099edccc3e506cf5336e8e2e0, torch 2.13.0+cpu, ibm-deeptools 2254.3a98611, ibm-flex 481.bf81df5, ibm-senlib 237.6e30780.
